packages/generate-persisted-query-manifest/src/index.ts

```diff
@@ -67,70 +71,134 @@ export const defaults: DefaultPersistedQueryManifestConfig = {
+  },
+  uniqueOperation: (name: string, source: DocumentSource) => {
+    return `Operation named "${COLORS.name(
+      name,
+    )}" already defined in: ${COLORS.filepath(source.file.path)}`;
+  },
+};
+
+function addError(source: DocumentSource, message: string) {
+  const vfileMessage = source.file.message(message, source.location);
+  vfileMessage.fatal = true;
+}
+
+function getDocumentSources(filepath: string): DocumentSource[] {
+  const file = vfile({
+    path: filepath,
+    contents: readFileSync(filepath, "utf-8"),
+  });
+
+  if (file.extname === ".graphql" || file.extname === ".gql") {
+    return [
+      {
+        node: parse(file.toString()),
+        file,
+        location: { line: 1, column: 1 },
+      },
+    ];
+  }
+
+  return gqlPluckFromCodeStringSync(filepath, file.toString()).map(
+    (source) => ({
+      node: parse(source.body),
+      file,
+      location: source.locationOffset,
+    }),
+  );
+}
+
 export async function generatePersistedQueryManifest(
   config: PersistedQueryManifestConfig = {},
 ): Promise<PersistedQueryManifest> {
-  const files = new Set<string>();
   const paths = config.documents ?? defaults.documents;
   const ignorePaths =
     config.documentIgnorePatterns ?? defaults.documentIgnorePatterns;
 
-  for (const f of await glob(paths, { ignore: ignorePaths })) {
-    files.add(f);
-  }
+  const filepaths = await glob(paths, { ignore: ignorePaths });
+  const sources = [...new Set(filepaths)].flatMap(getDocumentSources);
 
-  const documents = [...files].flatMap((filePath) => {
-    const content = readFileSync(filePath, "utf-8");
-    if (filePath.endsWith("graphql")) {
-      return [{ ast: parse(content), filename: filePath }];
-    } else {
-      return gqlPluckFromCodeStringSync(
-        filePath,
-        readFileSync(filePath, "utf-8"),
-      ).map((source) => ({ ast: parse(source.body), filename: source.name }));
-    }
-  });
+  const fragmentsByName = new Map<string, DocumentSource[]>();
+  const operationsByName = new Map<string, DocumentSource[]>();
 
-  interface LocatedOperationDefinitionNode {
-    node: OperationDefinitionNode;
-    filename: string;
-  }
+  for (const source of sources) {
+    visit(source.node, {
+      FragmentDefinition(node) {
+        const name = node.name.value;
+        const sources = fragmentsByName.get(name) ?? [];
 
-  const seenFragmentNames = new Set<string>();
-  const operationsByName = new Map<string, LocatedOperationDefinitionNode>();
-
-  for (const document of documents) {
-    for (const definition of document.ast.definitions) {
-      if (definition.kind === "FragmentDefinition") {
-        const name = definition.name.value;
-        if (seenFragmentNames.has(name)) {
-          throw new Error(`Duplicate fragment name: ${name}`);
+        if (sources.length) {
+          sources.forEach((sibling) => {
+            addError(source, ERROR_MESSAGES.uniqueFragment(name, sibling));
```

Comment:
It seems like this will add `N^2 - N` errors for any fragment that shows up N times. That seems pretty excessive? Can we just add a single error that lists the other N-1 places?

Reply:
I initially went that route, but found it a bit more difficult to reason about for a couple reasons:

1. Whichever file was "first" would get the error message. As the CLI, its difficult to determine which of the files is the "correct" file for the operation name so I felt it appropriate to report the error for all files involved.
2. I felt the formatting with multiple paths in the same message were difficult to parse.

I could be persuaded that combining paths for 2) is better than splitting out the error message. Here is the difference in how it looks:

**Current**
<img width="942" alt="Screenshot 2023-06-22 at 5 08 25 PM" src="https://github.com/apollographql/apollo-utils/assets/565661/5281202c-ca56-4bc5-b42d-5c05586fcdb1">

**Combined**
<img width="749" alt="Screenshot 2023-06-22 at 5 08 16 PM" src="https://github.com/apollographql/apollo-utils/assets/565661/c4b43d00-736c-4460-9067-eb64a89a36f2">

At minimum, I would print the error message for both files, which is how the [`@graphql-eslint/eslint-plugin`](https://the-guild.dev/graphql/eslint) packages prints these errors for both the [`unique-fragment-name`](https://the-guild.dev/graphql/eslint/rules/unique-fragment-name) rule and the [`unique-operation-name`](https://the-guild.dev/graphql/eslint/rules/unique-operation-name) rule.

Here is what that looks like in my editor for the eslint plugin:

<img width="724" alt="Screenshot 2023-06-22 at 5 00 03 PM" src="https://github.com/apollographql/apollo-utils/assets/565661/3f5484f5-01df-4267-9254-ec1ef516a59d">

Reply:
I suppose there's not that much of a difference between "O(N^2) error messages" and "O(N) error messages of length O(N) each", and I see your point that anchoring the error to each potentially relevant place does make sense!